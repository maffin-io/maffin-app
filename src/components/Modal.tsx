import React from 'react';

export type ModalProps = {
  open: boolean;
  setOpen: Function;
  title?: string;
  children: React.ReactNode;
};

export default function Modal({
  open, setOpen, title, children,
}: ModalProps): JSX.Element {
  if (!open) {
    return <span />;
  }

  return (
    <div
      className="fixed inset-0 bg-white bg-opacity-30 overflow-y-auto h-full w-full z-50"
      id="my-modal"
    >
      <div
        className="relative top-20 mx-auto p-5 w-1/3 shadow-lg rounded-md bg-gunmetal-700"
      >
        <button
          type="button"
          className="float-right"
          onClick={() => setOpen(false)}
        >
          X
        </button>
        {
          title
          && (
            <div className="w-full pb-4 border-b border-gunmetal-600">
              <span className="text-lg">
                {title}
              </span>
            </div>
          )
        }
        {children}
      </div>
    </div>
  );
}
