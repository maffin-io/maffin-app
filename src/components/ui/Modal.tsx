import React from 'react';
import RModal from 'react-modal';

export type ModalProps = {
  className?: string;
  triggerContent: string | JSX.Element;
  triggerClassName?: string;
  showClose?: boolean;
} & React.PropsWithChildren;

export interface ModalRef {
  openModal: () => void;
  closeModal: () => void;
}

function Modal({
  className = 'modal',
  triggerClassName = '',
  triggerContent,
  showClose = false,
  children,
}: ModalProps, ref: React.Ref<ModalRef>): JSX.Element {
  const [isOpen, setIsOpen] = React.useState<boolean>(false);

  // Expose actions to the parent
  React.useImperativeHandle(ref, () => ({
    openModal: () => setIsOpen(true),
    closeModal: () => setIsOpen(false),
  }));

  return (
    <>
      <RModal
        isOpen={isOpen}
        overlayClassName="overlay"
        className={className}
        onRequestClose={() => setIsOpen(false)}
      >
        {
          showClose
          && (
            <button
              type="button"
              className="float-right"
              onClick={() => setIsOpen(false)}
            >
              X
            </button>
          )
        }
        {children}
      </RModal>

      <button
        type="button"
        onClick={() => setIsOpen(!isOpen)}
        className={triggerClassName}
      >
        {triggerContent}
      </button>
    </>
  );
}

export default React.forwardRef(Modal);
