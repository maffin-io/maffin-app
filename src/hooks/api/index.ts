import { DateTime } from 'luxon';
import { mutate, SWRResponse } from 'swr';
import useSWRImmutable from 'swr/immutable';

import { Commodity } from '@/book/entities';
import { InvestmentAccount } from '@/book/models';
import * as queries from '@/lib/queries';
import type { PriceDBMap } from '@/book/prices';
import type { AccountsMap } from '@/types/book';
import type { Account, Split, Transaction } from '@/book/entities';

export function useStartDate(): SWRResponse<DateTime> {
  const key = '/api/start-date';
  return useSWRImmutable(
    key,
    fetcher(queries.getEarliestDate, key),
  );
}

export function useMainCurrency(): SWRResponse<Commodity> {
  const key = '/api/main-currency';
  return useSWRImmutable(
    key,
    fetcher(queries.getMainCurrency, key),
  );
}

export function useCommodity(guid: string): SWRResponse<Commodity> {
  const key = `/api/commodities/${guid}`;

  const result = useSWRImmutable(
    key,
    fetcher(() => Commodity.findOneByOrFail({ guid }), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  return result;
}

export function useCommodities(): SWRResponse<Commodity[]> {
  const key = '/api/commodities';

  const result = useSWRImmutable(
    key,
    fetcher(() => Commodity.find(), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  if (result.data) {
    result.data.forEach(
      (commodity: Commodity) => mutate(
        `/api/commodities/${commodity.guid}`,
        commodity,
        { revalidate: false },
      ),
    );
  }

  return result;
}

export function useLatestTxs(): SWRResponse<Transaction[]> {
  const key = '/api/txs/latest';
  return useSWRImmutable(
    key,
    fetcher(queries.getLatestTxs, key),
  );
}

export function useAccount(guid: string): SWRResponse<Account> {
  const key = `/api/accounts/${guid}`;

  const result = useSWRImmutable(
    key,
    fetcher(() => queries.getAccounts(guid), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  return result;
}

export function useAccounts(): SWRResponse<AccountsMap> {
  const key = '/api/accounts';

  const result = useSWRImmutable(
    key,
    fetcher(() => queries.getAccounts(), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  if (result.data) {
    Object.values(result.data as AccountsMap).forEach(
      (account: Account) => mutate(`/api/accounts/${account.guid}`, account, { revalidate: false }),
    );
  }

  return result;
}

export function useAccountsMonthlyTotals(): SWRResponse<queries.MonthlyTotals> {
  const { data: accounts } = useAccounts();
  const { data: prices } = useSWRImmutable(
    '/api/prices',
    fetcher(() => queries.getPrices({}), '/api/prices'),
  );

  const key = '/api/monthly-totals';
  const result = useSWRImmutable(
    (accounts && prices) ? key : null,
    fetcher(
      () => queries.getMonthlyTotals(accounts as AccountsMap, prices),
      key,
    ),
  );

  return result;
}

export function useInvestment(guid: string): SWRResponse<InvestmentAccount> {
  const key = `/api/investments/${guid}`;

  const result = useSWRImmutable(
    key,
    fetcher(() => queries.getInvestment(guid), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  return result;
}

export function useInvestments(): SWRResponse<InvestmentAccount[]> {
  const key = '/api/investments';

  const result = useSWRImmutable(
    key,
    fetcher(() => queries.getInvestments(), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  if (result.data) {
    result.data.forEach(
      (investment: InvestmentAccount) => mutate(
        `/api/investments/${investment.account.guid}`,
        investment,
        { revalidate: false },
      ),
    );
  }

  return result;
}

export function useSplits(guid: string): SWRResponse<Split[]> {
  const key = `/api/splits/${guid}`;
  return useSWRImmutable(
    key,
    fetcher(async () => queries.getSplits(guid), key),
  );
}

/**
 * Returns prices for a given commodity
 */
export function usePrices(c: Commodity | undefined): SWRResponse<PriceDBMap> {
  const key = `/api/prices/${c?.guid}`;
  return useSWRImmutable(
    c?.guid ? key : null,
    fetcher(async () => queries.getPrices({ from: c }), key),
  );
}

function fetcher(f: () => Promise<any>, key: string) {
  return async () => {
    const start = performance.now();
    const r = await f();
    const end = performance.now();
    console.log(`${key}: ${end - start}ms`);
    return r;
  };
}
