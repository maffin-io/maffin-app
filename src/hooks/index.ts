export { default as useDataSource, DataSourceContext } from './useDataSource';
export type { DataSourceContextType } from './useDataSource';
export { default as useBookStorage } from './useBookStorage';
export { default as useGapiClient } from './useGapiClient';
