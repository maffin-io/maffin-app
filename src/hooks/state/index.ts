export { useTheme } from './useTheme';
export { useOnline } from './useOnline';
