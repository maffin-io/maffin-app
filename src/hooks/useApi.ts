import useSWRImmutable from 'swr/immutable';
import { Fetcher, SWRResponse, SWRConfiguration } from 'swr';

import { Commodity } from '@/book/entities';
import * as queries from '@/lib/queries';
import { PriceDB } from '@/book/prices';
import getUser from '@/lib/getUser';

/**
 * Wrapper around SWR to make the usable keys explicit and re-usable
 */
export default function useApi(key: ApiPaths | null): SWRResponse {
  const f = key !== null ? API[key].f : () => {};
  const swrArgs = key !== null ? API[key].swrArgs : undefined;

  return useSWRImmutable(key, f, swrArgs);
}

export type ApiPaths = (
  '/api/start-date'
  | '/api/main-currency'
  | '/api/commodities'
  | '/api/accounts'
  | '/api/investments'
  | '/api/prices/today'
  | '/api/txs/latest'
  | '/api/user'
);

const API: {
  [key in ApiPaths]: {
    f: Fetcher,
    swrArgs?: SWRConfiguration,
  };
} = {
  '/api/start-date': {
    f: queries.getEarliestDate,
  },
  '/api/main-currency': {
    f: queries.getMainCurrency,
  },
  '/api/commodities': {
    f: Commodity.find,
  },
  '/api/accounts': {
    f: async (k: string) => {
      const start = performance.now();
      const r = await queries.getAccounts();
      const end = performance.now();
      console.log(`${k}: ${end - start}ms`);
      return r;
    },
  },
  '/api/investments': {
    f: queries.getInvestments,
  },
  '/api/prices/today': {
    f: PriceDB.getTodayQuotes,
  },
  '/api/txs/latest': {
    f: queries.getLatestTxs,
  },
  '/api/user': {
    f: getUser,
    swrArgs: {
      refreshInterval: 100000,
      revalidateOnMount: true,
    },
  },
};
