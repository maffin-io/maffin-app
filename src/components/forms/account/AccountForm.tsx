import React from 'react';
import { useForm, Controller } from 'react-hook-form';
import { classValidatorResolver } from '@hookform/resolvers/class-validator';
import { DateTime } from 'luxon';
import classNames from 'classnames';

import {
  Account,
  Commodity,
  Split,
  Transaction,
} from '@/book/entities';
import { Tooltip } from '@/components/tooltips';
import {
  AccountSelector,
  CommoditySelector,
  AccountTypeSelector,
} from '@/components/selectors';
import { useMainCurrency } from '@/hooks/api';
import { getAllowedSubAccounts } from '@/book/helpers/accountType';
import { toAmountWithScale } from '@/helpers/number';
import createEquityAccount from '@/lib/createEquityAccount';
import type { FormValues } from '@/components/forms/account/types';

const resolver = classValidatorResolver(Account, { validator: { stopAtFirstError: true } });

export type AccountFormProps =
| {
  action?: 'add',
  onSave?: Function,
  defaultValues?: Partial<FormValues>,
  hideDefaults?: boolean,
}
| {
  action: 'update' | 'delete',
  onSave?: Function,
  defaultValues?: Partial<FormValues>,
  hideDefaults?: never,
};

export default function AccountForm({
  action = 'add',
  defaultValues = {},
  onSave = () => {},
  hideDefaults = false,
}: AccountFormProps): React.JSX.Element {
  const { data: mainCurrency } = useMainCurrency();
  defaultValues.fk_commodity = defaultValues?.fk_commodity || mainCurrency;
  const form = useForm<FormValues>({
    defaultValues,
    mode: 'onChange',
    resolver,
  });

  const { errors } = form.formState;
  const disabled = action === 'delete';

  const parent = form.watch('parent');
  const ignoreTypes = (
    parent
    && Account.TYPES.filter(type => !getAllowedSubAccounts(parent.type).includes(type))
  ) || [];
  const type = form.watch('type');

  return (
    <form onSubmit={form.handleSubmit((data) => onSubmit(data, action, onSave))}>
      <div className="grid grid-cols-12 text-sm my-5 gap-2">
        <fieldset className="col-span-6">
          <label htmlFor="nameInput" className="inline-block mb-2">Name</label>
          <input
            id="nameInput"
            disabled={disabled}
            className="w-full m-0"
            {...form.register('name')}
            type="text"
          />
          <p className="invalid-feedback">{errors.name?.message}</p>
        </fieldset>

        <fieldset
          className={classNames(
            'col-start-8 col-span-2',
            {
              hidden: action === 'add',
            },
          )}
        >
          <label htmlFor="hiddenInput" className="inline-block mb-2">Hide</label>
          <span
            className="badge ml-0.5"
            data-tooltip-id="hidden-help"
          >
            ?
          </span>
          <Tooltip
            id="hidden-help"
          >
            <p>
              For accounts that are closed or not used anymore.
            </p>
          </Tooltip>
          <input
            id="hiddenInput"
            disabled={disabled}
            className="block m-0"
            {...form.register('hidden')}
            type="checkbox"
          />
        </fieldset>

        <fieldset
          className={classNames(
            'col-start-10 col-span-2',
            {
              hidden: hideDefaults && 'placeholder' in defaultValues,
            },
          )}
        >
          <label htmlFor="placeholderInput" className="inline-block mb-2">Parent</label>
          <span
            className="badge default ml-0.5"
            data-tooltip-id="placeholder-help"
          >
            ?
          </span>
          <Tooltip
            id="placeholder-help"
          >
            <p>
              For accounts that are parent categories. These accounts cannot have
              transactions.
            </p>
          </Tooltip>
          <input
            id="placeholderInput"
            disabled={disabled}
            className="block m-0"
            {...form.register('placeholder')}
            type="checkbox"
          />
        </fieldset>
      </div>
      <p className="invalid-feedback">{errors.placeholder?.message}</p>

      <fieldset
        className={classNames(
          'text-sm my-5',
          {
            hidden: hideDefaults && defaultValues?.parent,
          },
        )}
      >
        <label htmlFor="parent" className="inline-block mb-2">Parent</label>
        <Controller
          control={form.control}
          name="parent"
          render={({ field, fieldState }) => (
            <>
              <AccountSelector
                id="parentInput"
                isDisabled={disabled}
                isClearable={false}
                onlyPlaceholders
                placeholder="<parent account>"
                ignoreAccounts={[form.getValues('guid')]}
                onChange={field.onChange}
                defaultValue={defaultValues?.parent}
              />
              <p className="invalid-feedback">{fieldState.error?.message}</p>
            </>
          )}
        />
      </fieldset>

      <fieldset
        className={classNames(
          'text-sm my-5',
          {
            hidden: hideDefaults && defaultValues?.type,
          },
        )}
      >
        <label htmlFor="type" className="inline-block mb-2">Account type</label>
        <Controller
          control={form.control}
          name="type"
          render={({ field, fieldState }) => (
            <>
              <AccountTypeSelector
                id="typeInput"
                placeholder={parent ? '<select account type>' : '<select parent first>'}
                isDisabled={disabled || !parent}
                ignoreTypes={ignoreTypes}
                onChange={field.onChange}
                defaultValue={(defaultValues?.type && { type: defaultValues.type }) || undefined}
              />
              <p className="invalid-feedback">{fieldState.error?.message}</p>
            </>
          )}
        />
      </fieldset>

      <fieldset
        className={classNames(
          'grid grid-cols-12 text-sm my-5',
          {
            hidden: type !== 'BANK' || action !== 'add',
          },
        )}
      >
        <div
          className="col-span-6"
        >
          <label htmlFor="balanceInput" className="inline-block mb-2">Opening balance</label>
          <span
            className="badge default ml-0.5"
            data-tooltip-id="balance-help"
          >
            ?
          </span>
          <Tooltip
            id="balance-help"
          >
            <p>
              The initial amount you have in the account for the date
              you want to start tracking your finances.
            </p>
          </Tooltip>
          <input
            id="balanceInput"
            disabled={disabled}
            className="block w-full h-[38px] m-0 rounded-r-none"
            {...form.register(
              'balance',
              {
                valueAsNumber: true,
              },
            )}
            step="0.001"
            type="number"
          />
          <p className="invalid-feedback">{errors.balance?.message}</p>
        </div>
        <div
          className="col-span-6"
        >
          <label htmlFor="dateInput" className="inline-block mb-2">When</label>
          <span
            className="badge ml-0.5"
            data-tooltip-id="date-help"
          >
            ?
          </span>
          <Tooltip
            id="date-help"
          >
            <p>
              The date for the opening balance you&apos;ve selected.
            </p>
          </Tooltip>
          <input
            id="dateInput"
            className="block w-full m-0 rounded-l-none"
            {...form.register('balanceDate')}
            type="date"
            defaultValue={DateTime.now().toISODate() as string}
          />
          <p className="invalid-feedback">{errors.balanceDate?.message}</p>
        </div>
      </fieldset>

      <fieldset
        className={classNames(
          'text-sm my-5',
          {
            hidden: (hideDefaults && defaultValues?.fk_commodity)
                      || type === undefined
                      || ['INCOME', 'EXPENSE'].includes(type),
          },
        )}
      >
        <label htmlFor="commodity" className="inline-block mb-2">Commodity</label>
        <Controller
          control={form.control}
          name="fk_commodity"
          render={({ field, fieldState }) => (
            <>
              <CommoditySelector
                id="commodityInput"
                placeholder="Choose a commodity for the account"
                onChange={field.onChange}
                defaultValue={defaultValues?.fk_commodity as Commodity}
                isDisabled={disabled || action !== 'add'}
              />
              <p className="invalid-feedback">{fieldState.error?.message}</p>
            </>
          )}
        />
      </fieldset>

      <fieldset className="text-sm my-5">
        <label htmlFor="descriptionInput" className="inline-block mb-2">Description</label>
        <input
          id="descriptionInput"
          disabled={disabled}
          className="block w-full m-0"
          {...form.register('description')}
          type="text"
        />
        <p className="invalid-feedback">{errors.description?.message}</p>
      </fieldset>

      <input
        {...form.register(
          'guid',
        )}
        hidden
        type="text"
      />

      <div className="flex w-full justify-center">
        <button
          className={classNames(
            'btn capitalize',
            {
              'btn-primary': action === 'add',
              'btn-warn': action === 'update',
              'btn-danger': action === 'delete',
            },
          )}
          type="submit"
          disabled={Object.keys(errors).length > 0}
        >
          {action}
        </button>
      </div>
    </form>
  );
}

async function onSubmit(
  data: FormValues,
  action: 'add' | 'update' | 'delete',
  onSave: Function,
) {
  const account = Account.create({
    ...data,
    guid: data.guid || undefined,
  });

  if (action === 'add' || action === 'update') {
    await account.save();
    if (action === 'add' && data.balance) {
      await createBalance(data, account);
    }
  } else if (action === 'delete') {
    await account.remove();
  }

  onSave(account);
}

async function createBalance(
  data: FormValues,
  account: Account,
) {
  let equityAccount = await Account.findOneBy({
    type: 'EQUITY',
    name: `Opening balances - ${account.commodity.mnemonic}`,
  });

  if (!equityAccount) {
    equityAccount = await createEquityAccount(account.commodity);
  }

  const { amount, scale } = toAmountWithScale(data.balance as number);
  const denom = parseInt('1'.padEnd(scale + 1, '0'), 10);

  await Transaction.create({
    description: 'Opening balance',
    fk_currency: data.fk_commodity,
    splits: [
      Split.create({
        fk_account: account,
        valueNum: amount,
        valueDenom: denom,
        quantityNum: amount,
        quantityDenom: denom,
      }),
      Split.create({
        fk_account: equityAccount,
        valueNum: -amount,
        valueDenom: denom,
        quantityNum: -amount,
        quantityDenom: denom,
      }),
    ],
    date: data.balanceDate ? DateTime.fromISO(data.balanceDate as string) : DateTime.now(),
  }).save();
}
