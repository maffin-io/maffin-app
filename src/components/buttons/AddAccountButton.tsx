import React from 'react';
import Modal from '@/components/Modal';
import { BiPlusCircle } from 'react-icons/bi';

import { useDataSource } from '@/hooks';
import AccountForm from '@/components/forms/account/AccountForm';

export default function AddAccountButton(): JSX.Element {
  const { save } = useDataSource();
  const [isModalOpen, setIsModalOpen] = React.useState<boolean>(false);

  return (
    <>
      <Modal
        title="Add account"
        open={isModalOpen}
        setOpen={setIsModalOpen}
      >
        <AccountForm
          onSave={() => {
            save();
            setIsModalOpen(false);
          }}
        />
      </Modal>
      <button
        type="button"
        className="btn-primary"
        onClick={() => setIsModalOpen(!isModalOpen)}
      >
        <BiPlusCircle className="mr-1" />
        Add Account
      </button>
    </>
  );
}
