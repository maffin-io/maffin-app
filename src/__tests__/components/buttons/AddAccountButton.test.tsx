import React from 'react';
import {
  act,
  render,
  screen,
  fireEvent,
} from '@testing-library/react';
import Modal from 'react-modal';

import AddAccountButton from '@/components/buttons/AddAccountButton';
import AccountForm from '@/components/forms/account/AccountForm';
import { DataSourceContext } from '@/hooks';
import type { DataSourceContextType } from '@/hooks';

jest.mock('react-modal', () => jest.fn(
  (props: React.PropsWithChildren) => (
    <div data-testid="Modal">
      {props.children}
    </div>
  ),
));

jest.mock('@/components/forms/account/AccountForm', () => jest.fn(
  () => <div data-testid="AccountForm" />,
).mockName('AccountForm'));
const AccountFormMock = AccountForm as jest.MockedFunction<typeof AccountForm>;

describe('AddAccountButton', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders hidden modal on mount', async () => {
    render(<AddAccountButton />);
    expect(Modal).toHaveBeenLastCalledWith(
      expect.objectContaining({
        isOpen: false,
        className: 'relative top-20 mx-auto p-5 w-1/3 shadow-lg rounded-md bg-gunmetal-700',
        overlayClassName: 'fixed inset-0 bg-white bg-opacity-30 overflow-y-auto h-full w-full z-50',
      }),
      {},
    );

    expect(AccountForm).toHaveBeenLastCalledWith(
      {
        onSave: expect.any(Function),
      },
      {},
    );
  });

  it('closes modal and saves', async () => {
    const mockSave = jest.fn();
    render(
      <DataSourceContext.Provider value={{ save: mockSave as Function } as DataSourceContextType}>
        <AddAccountButton />
      </DataSourceContext.Provider>,
    );

    // open modal to prove that onSave closes it
    const button = await screen.findByRole('button', { name: /add account/i });
    fireEvent.click(button);
    expect(Modal).toHaveBeenLastCalledWith(
      expect.objectContaining({
        isOpen: true,
        className: 'relative top-20 mx-auto p-5 w-1/3 shadow-lg rounded-md bg-gunmetal-700',
        overlayClassName: 'fixed inset-0 bg-white bg-opacity-30 overflow-y-auto h-full w-full z-50',
      }),
      {},
    );

    const { onSave } = AccountFormMock.mock.calls[0][0];
    act(() => onSave());
    expect(mockSave).toBeCalledTimes(1);

    expect(Modal).toHaveBeenLastCalledWith(
      expect.objectContaining({
        isOpen: false,
        className: 'relative top-20 mx-auto p-5 w-1/3 shadow-lg rounded-md bg-gunmetal-700',
        overlayClassName: 'fixed inset-0 bg-white bg-opacity-30 overflow-y-auto h-full w-full z-50',
      }),
      {},
    );
  });

  it('opens modal when clicking the button', async () => {
    render(
      <AddAccountButton />,
    );

    const button = await screen.findByRole('button', { name: /add account/i });
    fireEvent.click(button);

    const modal = await screen.findByTestId('Modal');
    expect(modal).toMatchSnapshot();
  });
});
