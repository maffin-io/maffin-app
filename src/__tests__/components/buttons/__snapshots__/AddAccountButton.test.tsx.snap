// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddAccountButton opens modal when clicking the button 1`] = `
<div
  data-testid="Modal"
>
  <button
    class="float-right"
    type="button"
  >
    X
  </button>
  <span>
    Add account
  </span>
  <div
    data-testid="AccountForm"
  />
</div>
`;
